libnmstate/state.py

```diff
@@ -0,0 +1,248 @@
+#
+# Copyright 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+import collections
+import copy
+from operator import itemgetter
+import six
+
+from libnmstate import iplib
+from libnmstate import metadata
+from libnmstate.error import NmstateVerificationError
+from libnmstate.prettystate import format_desired_current_state_diff
+from libnmstate.schema import Interface
+from libnmstate.schema import InterfaceType
+
+
+def create_state(state, interfaces_to_filter=None):
+    """
+    Create a state object, given an initial state.
+    interface_filter: Limit the interfaces included in the state to the ones
+    mentioned in the list. None implied no filtering.
+    """
+    new_state = {}
+    if interfaces_to_filter is not None:
+        origin = State(state)
+        iface_names = set(origin.interfaces) & interfaces_to_filter
+        filtered_ifaces_state = [
+            origin.interfaces[ifname] for ifname in iface_names
+        ]
+        new_state[Interface.KEY] = filtered_ifaces_state
+
+    return State(new_state)
+
+
+class State(object):
+    def __init__(self, state):
+        self._state = copy.deepcopy(state)
+        self._ifaces_state = State._index_interfaces_state_by_name(self._state)
+
+    def __eq__(self, other):
+        return self.state == other.state
+
+    def __hash__(self):
+        return hash(self.state)
+
+    def __str__(self):
+        return str(self.state)
+
+    def __repr__(self):
+        return self.__str__()
+
+    @property
+    def state(self):
+        self._state[Interface.KEY] = sorted(
+            list(six.viewvalues(self._ifaces_state)),
+            key=itemgetter('name')
+        )
+        return self._state
+
+    @property
+    def interfaces(self):
+        """ Indexed interfaces state """
+        return self._ifaces_state
+
+    def sanitize_ethernet(self, other_state):
```

Comment:
I was wondering if it makes sense to actually store both the current full state and the desired (possibly partial) state and then let the NM drivers access both and merge the state as required. Then they can directly see what was explicitly stated by the user and decide what to do. The special handling can then be done in the actual driver.

Reply:
I think that the approach to push down the stack the merging has been attempted or discussed already (not sure).
If you go in this path, we will need a translator (between nmstate API and NM one) and will need to repeat this merging on each provider (assuming more will come in).
The question is, does it help differing the merge further in, removing the common code that does the merging.

I think that the sanitize step can be just removed by avoiding the problem of RW parameter side effects, like we did in the routing case. Whenever we have an issue that one RW parameter affects other RW parameters, we should split between configuration and running states.
`ethernet` and `ip-address` are good examples.

Reply:
So your idea is to have

```
ethernet:
    config: ...
    running: ...
```

and the same for IP addresses? This seems like an ugly solution. Will it be possible to change the running state in this example or only the config state?

Regarding the common code for merging, we can still provide common code to get the merged results if this is useful and provide access to the unmerged states.

Also I would only care about other provides when they are there instead of complicating things for events that might never happen.